src/tracer.c

```diff
@@ -195,42 +147,89 @@ handle_syscall(pid_t pid, const struct ptrace_syscall_info *entry,
 	return;			       // return on syscall failure
     }
 
-    const int syscall = entry->entry.nr;
-
-    if (syscall != SYS_open && syscall != SYS_creat && syscall != SYS_openat
-	&& syscall != SYS_close) {
-	return;			       // return if we don't care about
-	// tracking said syscall
-    }
-
-    if (syscall == SYS_close) {
-	handle_close(pinfo->finfo + entry->entry.args[0]);
-	return;
-    }
-
-    const int fd = exit->exit.rval;
+    int fd;
+    void *path;
+    int flags;
+    int dirfd;
 
-    PROCESS_INFO *pinfo = find(pid);
-    FILE_INFO *finfo;
+    switch (entry->entry.nr) {
+	case SYS_open:
+	    // int open(const char *pathname, int flags, ...);
+	    fd = (int) exit->exit.rval;
+	    path = (void *) entry->entry.args[0];
+	    flags = (int) entry->entry.args[1];
 
-    if (fd == pinfo->finfo_size) {
-	finfo = next_finfo(pinfo);
-    } else {
-	finfo = pinfo->finfo + fd;
-    }
+	    FILE_INFO *finfo = finfo_at(find(pid), fd);
 
-    switch (syscall) {
-	case SYS_open:
-	    handle_open(pid, finfo, entry->entry.args);
+	    finfo->path = get_str_from_process(pid, path);
+	    finfo->purpose = flags;
 	    break;
 	case SYS_creat:
-	    handle_creat(pid, finfo, entry->entry.args);
+	    // int creat(const char *pathname, ...);
+	    fd = (int) exit->exit.rval;
+	    path = (void *) entry->entry.args[0];
+
+	    FILE_INFO *finfo = finfo_at(find(pid), fd);
+
+	    finfo->path = get_str_from_process(pid, path);
+	    finfo->purpose = O_CREAT | O_WRONLY | O_TRUNC;
 	    break;
 	case SYS_openat:
-	    handle_openat(pid, finfo, entry->entry.args);
+	    // int openat(int dirfd, const char *pathname, int flags, ...);
+	    fd = (int) exit->exit.rval;
+	    dirfd = (int) entry->entry.args[0];
+	    path = (void *) entry->entry.args[1];
+	    flags = (int) entry->entry.args[2];
+
+	    FILE_INFO *finfo = finfo_at(find(pid), fd);
+	    char *rpath = get_str_from_process(pid, path);
+
+	    finfo->purpose = flags;
+
+	    if (dirfd == AT_FDCWD || *rpath == '/') {	// If it's an
+		// absolute path or
```

Comment:
comment on line below, to help indent

Reply:
Oh i didn't see that. I will add to it when i get back home.

Reply:
@zvr done